Index: hostapd-2022-09-19-01944c09/src/ap/hostapd.c
===================================================================
--- hostapd-2022-09-19-01944c09.orig/src/ap/hostapd.c
+++ hostapd-2022-09-19-01944c09/src/ap/hostapd.c
@@ -4275,7 +4275,7 @@ static int hostapd_change_config_freq(st
 				      struct hostapd_freq_params *old_params)
 {
 	int channel;
-	u8 seg0, seg1 = 0;
+	u8 seg0 = 0, seg1 = 0;
 	struct hostapd_hw_modes *mode;
 
 	if (!params->channel) {
