Index: hostapd-2022-09-19-01944c09/src/drivers/driver.h
===================================================================
--- hostapd-2022-09-19-01944c09.orig/src/drivers/driver.h
+++ hostapd-2022-09-19-01944c09/src/drivers/driver.h
@@ -2312,6 +2312,9 @@ struct wpa_driver_capa {
 	/** Maximum number of iterations in a single scan plan */
 	u32 max_sched_scan_plan_iterations;
 
+	/** Maximum number of extra IE bytes for scans */
+	u16 max_scan_ie_len;
+
 	/** Whether sched_scan (offloaded scanning) is supported */
 	int sched_scan_supported;
 
Index: hostapd-2022-09-19-01944c09/src/drivers/driver_nl80211_capa.c
===================================================================
--- hostapd-2022-09-19-01944c09.orig/src/drivers/driver_nl80211_capa.c
+++ hostapd-2022-09-19-01944c09/src/drivers/driver_nl80211_capa.c
@@ -952,6 +952,10 @@ static int wiphy_info_handler(struct nl_
 			nla_get_u32(tb[NL80211_ATTR_MAX_SCAN_PLAN_ITERATIONS]);
 	}
 
+	if (tb[NL80211_ATTR_MAX_SCAN_IE_LEN])
+		capa->max_scan_ie_len =
+			nla_get_u16(tb[NL80211_ATTR_MAX_SCAN_IE_LEN]);
+
 	if (tb[NL80211_ATTR_MAX_MATCH_SETS])
 		capa->max_match_sets =
 			nla_get_u8(tb[NL80211_ATTR_MAX_MATCH_SETS]);
Index: hostapd-2022-09-19-01944c09/src/drivers/driver_nl80211_scan.c
===================================================================
--- hostapd-2022-09-19-01944c09.orig/src/drivers/driver_nl80211_scan.c
+++ hostapd-2022-09-19-01944c09/src/drivers/driver_nl80211_scan.c
@@ -222,7 +222,7 @@ nl80211_scan_common(struct i802_bss *bss
 		wpa_printf(MSG_DEBUG, "nl80211: Passive scan requested");
 	}
 
-	if (params->extra_ies) {
+	if (params->extra_ies && drv->capa.max_scan_ie_len >= params->extra_ies_len) {
 		wpa_hexdump(MSG_MSGDUMP, "nl80211: Scan extra IEs",
 			    params->extra_ies, params->extra_ies_len);
 		if (nla_put(msg, NL80211_ATTR_IE, params->extra_ies_len,
