diff --git a/drv/ppe_drv/tun/ppe_drv_tun_prgm_prsr.c b/drv/ppe_drv/tun/ppe_drv_tun_prgm_prsr.c
--- a/drv/ppe_drv/tun/ppe_drv_tun_prgm_prsr.c
+++ b/drv/ppe_drv/tun/ppe_drv_tun_prgm_prsr.c
@@ -469,7 +469,7 @@ static bool ppe_drv_tun_prgm_prsr_fixed_
 static bool ppe_drv_tun_prgm_prsr_cfg_equal(struct ppe_drv_tun_prgm_prsr_cfg *a,
 						struct ppe_drv_tun_prgm_prsr_cfg *b)
 {
-	struct ppe_drv *p = ppe_drv_gbl;
+	struct ppe_drv __maybe_unused *p = ppe_drv_gbl;

 	if (!a || !b) {
 		return false;
diff --git a/drv/ppe_drv/tun/ppe_drv_tun_udf.c b/drv/ppe_drv/tun/ppe_drv_tun_udf.c
--- a/drv/ppe_drv/tun/ppe_drv_tun_udf.c
+++ b/drv/ppe_drv/tun/ppe_drv_tun_udf.c
@@ -100,7 +100,7 @@
  */
 bool ppe_drv_tun_udf_compare_udf_profiles(struct ppe_drv_tun_udf_profile *a, struct ppe_drv_tun_udf_profile *b)
 {
-	struct ppe_drv *p = ppe_drv_gbl;
+	struct ppe_drv __maybe_unused *p = ppe_drv_gbl;
 	int i;

 	/*
