Index: u-boot-freedom-2023.04.02-prpl/arch/arm/dts/ipq9574-rdp433-wnc-freedom.dts
===================================================================
--- u-boot-freedom-2023.04.02-prpl.orig/arch/arm/dts/ipq9574-rdp433-wnc-freedom.dts
+++ u-boot-freedom-2023.04.02-prpl/arch/arm/dts/ipq9574-rdp433-wnc-freedom.dts
@@ -10,6 +10,10 @@
 / {
 	config_name = "config@al02-c4", "config@rdp433-wnc-freedom", "config-rdp433-wnc-freedom";
 
+	chosen {
+		u-boot,booted-bank="active";
+	};
+
 	soc: soc {
 		tlmm: pinctrl@1000000 {
 
